source/common/quic/envoy_tls_server_handshaker.cc

```diff
@@ -0,0 +1,115 @@
+namespace Quic {
+
+EnvoyTlsServerHandshaker::EnvoyTlsServerHandshaker(
+    quic::QuicSession* session, const quic::QuicCryptoServerConfig* crypto_config,
+    Envoy::Event::Dispatcher& dispatcher)
+    : quic::TlsServerHandshaker(session, crypto_config), dispatcher_(dispatcher),
+      proof_source_(crypto_config->proof_source()) {}
+
+std::unique_ptr<quic::ProofSourceHandle> EnvoyTlsServerHandshaker::MaybeCreateProofSourceHandle() {
+  return std::make_unique<EnvoyDefaultProofSourceHandle>(
+    this, proof_source_, quic::TlsServerHandshaker::MaybeCreateProofSourceHandle());
+}
+
+EnvoyTlsServerHandshaker::EnvoyDefaultProofSourceHandle::EnvoyDefaultProofSourceHandle(
+  EnvoyTlsServerHandshaker* tls_server_handshaker,
+  quic::ProofSource* proof_source,
+  std::unique_ptr<quic::ProofSourceHandle> default_proof_source_handle) :
+    handshaker_(tls_server_handshaker),
+    proof_source_(dynamic_cast<EnvoyQuicProofSource*>(proof_source)),
+    default_proof_source_handle_(std::move(default_proof_source_handle)) {
+  ASSERT(handshaker_ != nullptr);
+}
+
+EnvoyTlsServerHandshaker::EnvoyDefaultProofSourceHandle::~EnvoyDefaultProofSourceHandle() {
+  CloseHandle();
+}
+
+// Close the handle. Cancel the pending signature operation, if any.
+void EnvoyTlsServerHandshaker::EnvoyDefaultProofSourceHandle::CloseHandle() {
+  if (private_key_method_provider_ == nullptr) {
+    default_proof_source_handle_->CloseHandle();
+  }
+}
+
+// Delegates to proof_source_->GetCertChain.
+// Returns QUIC_SUCCESS or QUIC_FAILURE. Never returns QUIC_PENDING.
+quic::QuicAsyncStatus EnvoyTlsServerHandshaker::EnvoyDefaultProofSourceHandle::SelectCertificate(
+    const quic::QuicSocketAddress& server_address,
+    const quic::QuicSocketAddress& client_address,
+    const quic::QuicConnectionId& original_connection_id, absl::string_view ssl_capabilities,
+    const std::string& hostname, absl::string_view client_hello, const std::string& alpn,
+    std::optional<std::string> alps, const std::vector<uint8_t>& quic_transport_params,
+    const std::optional<std::vector<uint8_t>>& early_data_context,
+    const quic::QuicSSLConfig& ssl_config) {
+  return default_proof_source_handle_->SelectCertificate(
+    server_address, client_address, original_connection_id, ssl_capabilities,
+    hostname, client_hello, alpn, alps, quic_transport_params, early_data_context, ssl_config);
+}
+
+quic::QuicAsyncStatus EnvoyTlsServerHandshaker::EnvoyDefaultProofSourceHandle::ComputeSignature(
+    const quic::QuicSocketAddress& server_address, const quic::QuicSocketAddress& client_address,
+    const std::string& hostname, uint16_t signature_algorithm, absl::string_view in,
+    size_t max_signature_size) {
+  auto res = proof_source_->getTransportSocketAndFilterChain(server_address, client_address, hostname);
+  if (!res.has_value()) {
+    return quic::QUIC_FAILURE;
+  }
+
+  private_key_method_provider_ =
+      res->transport_socket_factory_.getPrivateKeyMethodProvider(hostname);
```

Comment:
Could you check getPrivateKeyMethodProvider() in MaybeCreateProofSourceHandle() and only create EnvoyDefaultProofSourceHandle if a private key method provider is configured, otherwise just call into the base class?

Reply:
I guess I can't, since I need the hostname. We don't have the hostname when invoke `MaybeCreateProofSourceHandle`

Reply:
Ah, I see. hostname isn't plumbed to MaybeCreateProofSourceHandle() though it is already available.